src/java/org/apache/cassandra/service/AbstractReadExecutor.java

```diff
@@ -157,6 +181,68 @@ public Row get() throws ReadFailureException, ReadTimeoutException, DigestMismat
         return handler.get();
     }
 
+    /**
+     * Compare difference between passed timestamp and start field against Threshold cutoffs. If the threshold is
+     * exceeded, the current p99 latency of the retry endpoint is added to the threshold as the "predicted performance"
+     */
+    public void writePredictedSpeculativeRetryPerformanceMetrics(long timestamp) {
+        InetAddress extraReplica = Iterables.getLast(targetReplicas);
+        for (Threshold threshold : Threshold.values()) {
+            writePredSpecRetryPerformanceMetricByThreshold(threshold, timestamp, extraReplica);
+        }
+    }
+
+    private void writePredSpecRetryPerformanceMetricByThreshold(Threshold threshold, long timestamp, InetAddress extraReplica) {
```

Comment:
Can we throw all of this into a separate class? Will be easier to manage future commits/patches.